### DISCUSSION TITLE
[SOLVED] RSS Guard can localize feeds

### DISCUSSION THREAD

**Discussion by martinrotter** [martinrotter/rssguard#477] [Q&A]
@StealerCook - if you are interested, ping me in this discussion.

**Comment by StealerCook** [comment]
Hey Hi, Is this about Feed translation? Sry i don't remember starting thread here...

**Reply by martinrotter** [reply]
Hello, no I just marked you to notice you.

I say that you asked about localizing feeds in quiterss - https://github.com/QuiteRSS/quiterss/issues/1475

Given the fact that quiterss is defacto dead - not developed anymore - I wanted to offer alternative.

RSS Guard does support localizing feeds via its "website scraping" [feature](https://github.com/martinrotter/rssguard/blob/master/resources/docs/Documentation.md#scrap).

You need to download and install Python for Windows, add "python.exe" into your `PATH`. Then you need to download the [script](https://github.com/martinrotter/rssguard/blob/master/resources/scripts/scrapers/translate-feed.py) which performs the localization of feed. Check the top of the script for more things you need to install to have the script working.

Then you employ the script via the above "websites scraping" feature (do not get fooled by the name, it does exactly what you want).

The script can localize any RSS/RDF/ATOM feed from any to any language.

**Reply by StealerCook** [reply]
Oh Thankyou Mate, This is great. I will try this out and revert back for any guidance.

**Reply by martinrotter** [reply]
Sure, no prob. It is relatively advanced functionality. Some knowledge is needed - installing Python, adding "exe" to "PATH" variable, launching executables on command line (to test the script) etc., but it is all needed because localization of feeds is not directly implemented as GUI option in any feed reader I know and only some feed readers offer reasonable "work-around" - the only I know are RSS Guard (of course), Liferea and probably newsboat.

Thing is that once you get the grasp of the "scraping" feature, you can do crazy magic like filtering out advertisements, CSS/HTML formatting or you can even generate feeds for any remote or local data source.

**Reply by martinrotter** [reply]
Here is a screenshot of localized feed I tested right now. Feed is localized from english to portuguese - notice the "Post-process script" box where the script is called with parameters "en" (english) "pt_BR" (portuguese):

![image](https://user-images.githubusercontent.com/1255302/131094328-d2baaffb-753c-451e-b796-e3c1fc2dcf44.png)

**Reply by martinrotter** [reply]
![image](https://user-images.githubusercontent.com/1255302/131094461-ef4f5d95-fd2c-4280-85c8-5dbdd98c850e.png)

**Reply by martinrotter** [reply]
Also, it is probably best if you use latest "development build" - https://github.com/martinrotter/rssguard/releases/tag/devbuild

**Reply by StealerCook** [reply]
Thankyou. I can manage with python settings and thanks for your guide. It really has cool features.

**Comment by StealerCook** [comment]
@martinrotter just tried this out. But I get this error. Did I miss anything?
![image](https://user-images.githubusercontent.com/76549252/134457644-7f27a144-458f-4eb2-9812-44fd584dfd7f.png)

**Reply by martinrotter** [reply]
@StealerCook what is `py -3` in the beginning of the post-process script line? it should be something like `C:\path\to\python.exe` or just `python` if it is in your `PATH`. Also, it is good to launch rssguard with `--log` switch (see `--help` for more info) and generate debug log and investigate the log as it may contain more info

**Comment by StealerCook** [comment]
Hey Thankyou. I thought it would work. I tried with the python path, now it isn't providing any error. But i don't see the translated feed. In the left side you can still see feeds in English.
![image](https://user-images.githubusercontent.com/76549252/134464711-d3c5991c-90f4-48e4-a03d-0e2e7623cd46.png)

**Reply by martinrotter** [reply]
what happens when you push "Fetch it now" button? do you get any error message?

**Reply by StealerCook** [reply]
No Error. Just shows "All metadata fetched successfully."

**Reply by martinrotter** [reply]
Just tested with your feed. I had to increase feed download timeout in settings, then:
![image](https://user-images.githubusercontent.com/1255302/134468120-398000c7-2938-4c36-9723-7d0435f317b9.png)

Notice that title of feed is now translated after "Fetch it now". Also, do you have all dependecies installed? They are listed in the script file - read it.

![image](https://user-images.githubusercontent.com/1255302/134468331-91cac17c-cd6b-4741-9a9a-f3b65364638a.png)

**Reply by martinrotter** [reply]
Also, you can test the script from command line to get better info if it actually works.

**Comment by StealerCook** [comment]
Yes, downloaded all the dependencies. How do you change download timeout, do you mean that feed connection timeout?
I tried with command line too and that provided me this.
![image](https://user-images.githubusercontent.com/76549252/134470396-4a8e4448-f12f-47af-8aab-9dae31de8ae3.png)

**Reply by martinrotter** [reply]
Yes, that's exactly how you should test it, nice. Can you test it the same way with latest script file?

https://github.com/martinrotter/rssguard/blob/master/resources/scripts/scrapers/translate-feed.py

**Reply by martinrotter** [reply]
> do you mean that feed connection timeout?

Yes but it seems it has nothing to do with your problems as there is a clear syntax error detected in XML file in script you used.

**Reply by martinrotter** [reply]
It fails on line 40, it fails to parse the XML, weird. https://github.com/martinrotter/rssguard/blob/master/resources/scripts/scrapers/translate-feed.py#L40

**Reply by martinrotter** [reply]
Maybe the ENCODING of your environment is NOT UTF-8 and you thus may need to pass additional parameter to the script. There 4th parametr "encoding", you can use like this:

`translate-feed.py en pt_BR true utf-8`

except you maybe need to use something else instead of utf-8?

if your run "chcp.exe" in your powershell/cmd what does it print?

**Reply by StealerCook** [reply]
chcp value is 65001 and I tried these with the pointed script only.
Tried with this >>translate-feed.py en pt_BR true utf-8
but command line error remains the same.
To note the errors are different.
![image](https://user-images.githubusercontent.com/76549252/134476293-c9fc2fc3-40a0-4df0-8743-1394e31b46cd.png)
![image](https://user-images.githubusercontent.com/76549252/134476360-d17465fe-2f39-4132-a3db-2f1b5d3e94d7.png)

**Reply by martinrotter** [reply]
it is better to just try to execute script in command line first and once that works, move to rss guard. "utf-8" is the default, you only need to specify encoding once it's different

i test now and also got error, let me test

**Reply by martinrotter** [reply]
btw, your screenshot from rssguard is promising, it failed at the end of script, 

also, you use powershell? if so then your command is wrong, it must be:
`
curl https://www.reddit.com/r/TOR.rss | Select-Object -ExpandProperty Content | python.exe .\translate-feed.py en pt true`

**Reply by martinrotter** [reply]
one more discovered thing, if you try many times, reddit returns HTML (not XML) saying that you are a robot etc

**Reply by martinrotter** [reply]
i uploaded new version of the script! please try it.

https://github.com/martinrotter/rssguard/blob/master/resources/scripts/scrapers/translate-feed.py

**Reply by StealerCook** [reply]
Thankyou for updating the script. I just ran below command in PowerShell. No Error. But doesn't translate.
>> curl https://www.reddit.com/r/TOR.rss | Select-Object -ExpandProperty Content | python feed_translate.py en pt true
![image](https://user-images.githubusercontent.com/76549252/134497670-e8657d6e-3229-413c-be55-cad73cd77a96.png)

**Reply by martinrotter** [reply]
ok, there is a function "translate_string" in

https://github.com/martinrotter/rssguard/blob/master/resources/scripts/scrapers/translate-feed.py#L58

tweak it so that it does not enclose its inner calls in try block:

```python
def translate_string(to_translate):
    if to_translate is None:
      return to_translate

    translated_text = translator.translate(to_translate, src = lang_from, dest = lang_to)

    if not parallel:
      time.sleep(0.2)

    return translated_text.text
```

then try again and you should see the actual errors



I will repeat one thing. **Those scripts in the repository are semi-DIY and I completely know that they will not work out-of-box for everyone and adjustments might have to be made.**
